File: src/app/core/ledger/run_record.py
from datetime import UTC, datetime
from typing import Any

BEAMPIPE_RUN_RECORD_KEY = "beampipe_run_record"
_RAW_LINE_MAX = 512
_GRAPH_ERROR_UIDS_MAX = 20


def _now_iso_z() -> str:
    return datetime.now(UTC).isoformat().replace("+00:00", "Z")


def _trim_raw(raw: str | None) -> str | None:
    if not raw:
        return None
    s = raw.strip()
    if len(s) <= _RAW_LINE_MAX:
        return s
    return s[: _RAW_LINE_MAX] + "..."


def _observation_dict(
    *,
    state: str,
    exit_code: Any,
    source: str,
    raw_line: str | None,
) -> dict[str, Any]:
    observed_at = _now_iso_z()
    out: dict[str, Any] = {
        "state": state,
        "exit_code": exit_code,
        "source": source,
        "observed_at": observed_at,
    }
    tr = _trim_raw(raw_line)
    if tr:
        out["raw"] = tr
    return out


def _merge_paths(slurm: dict[str, Any], paths: dict[str, str] | None) -> None:
    if not paths:
        return
    merged = dict(slurm.get("paths") or {})
    for k, v in paths.items():
        if v:
            merged[str(k)] = str(v).strip()
    if merged:
        slurm["paths"] = merged


def merge_slurm_poll_into_manifest(
    existing_manifest: dict[str, Any] | None,
    *,
    composite_scheduler_job_id: str,
    slurm_job_id: str,
    state: str,
    source: str,
    exit_code: Any,
    raw_line: str | None,
    record_terminal: bool,
    terminal_ledger_status: str | None,
    remote_session_dir: str | None = None,
    diagnostics: dict[str, Any] | None = None,
    reason: str | None = None,
) -> dict[str, Any]:
    base: dict[str, Any] = dict(existing_manifest) if isinstance(existing_manifest, dict) else {}
    rr: dict[str, Any] = dict(base.get(BEAMPIPE_RUN_RECORD_KEY) or {})
    slurm: dict[str, Any] = dict(rr.get("slurm") or {})
    slurm.pop("observations", None)

    slurm["slurm_job_id"] = str(slurm_job_id)
    slurm["composite_scheduler_job_id"] = str(composite_scheduler_job_id)

    obs = _observation_dict(
        state=state,
        exit_code=exit_code,
        source=source,
        raw_line=raw_line,
    )
    slurm["last_observation"] = obs

    if remote_session_dir and str(remote_session_dir).strip():
        _merge_paths(slurm, {"session_dir": str(remote_session_dir).strip()})

    if record_terminal and terminal_ledger_status:
        comp = str(composite_scheduler_job_id)
        existing_terminal = slurm.get("terminal")
        frozen = (
            isinstance(existing_terminal, dict)
            and existing_terminal.get("ledger_status") == terminal_ledger_status
            and str(existing_terminal.get("composite_scheduler_job_id") or "") == comp
        )
        if not frozen:
            term: dict[str, Any] = {
                **obs,
                "ledger_status": terminal_ledger_status,
                "composite_scheduler_job_id": comp,
                "slurm_job_id_ref": str(slurm_job_id),
            }
            if reason:
                term["reason"] = str(reason)
            if diagnostics:
                term["diagnostics"] = {k: v for k, v in diagnostics.items() if v is not None}
            slurm["terminal"] = term

    rr["slurm"] = slurm
    out = dict(base)
    out[BEAMPIPE_RUN_RECORD_KEY] = rr
    return out


def merge_slurm_submit_into_manifest(
    existing_manifest: dict[str, Any] | None,
    *,
    session_id: str,
    slurm_job_id: str,
    composite_scheduler_job_id: str,
    login_node: str | None,
    remote_user: str | None,
    paths: dict[str, str] | None = None,
) -> dict[str, Any]:
    """Stamp Slurm submit metadata (after sbatch)."""
    base: dict[str, Any] = dict(existing_manifest) if isinstance(existing_manifest, dict) else {}
    rr: dict[str, Any] = dict(base.get(BEAMPIPE_RUN_RECORD_KEY) or {})
    slurm: dict[str, Any] = dict(rr.get("slurm") or {})
    slurm.pop("observations", None)
    slurm["session_id"] = str(session_id)
    slurm["slurm_job_id"] = str(slurm_job_id)
    slurm["composite_scheduler_job_id"] = str(composite_scheduler_job_id)
    slurm["submitted_at"] = _now_iso_z()
    if login_node:
        slurm["login_node"] = str(login_node).strip()
    if remote_user:
        slurm["remote_user"] = str(remote_user).strip()
    _merge_paths(slurm, paths)
    rr["slurm"] = slurm
    out = dict(base)
    out[BEAMPIPE_RUN_RECORD_KEY] = rr
    return out


def merge_dim_deploy_into_manifest(
    existing_manifest: dict[str, Any] | None,
    *,
    session_id: str,
    dim_rest_base: str,
    verify_ssl: bool,
    operator_urls: dict[str, str] | None = None,
) -> dict[str, Any]:
    """Record successful DIM deploy."""
    base: dict[str, Any] = dict(existing_manifest) if isinstance(existing_manifest, dict) else {}
    rr: dict[str, Any] = dict(base.get(BEAMPIPE_RUN_RECORD_KEY) or {})
    dim: dict[str, Any] = dict(rr.get("dim") or {})
    dim.pop("observations", None)
    dim["session_id"] = str(session_id)
    deploy: dict[str, Any] = dict(dim.get("deploy") or {})
    deploy["deployed_at"] = _now_iso_z()
    deploy["dim_rest_base"] = str(dim_rest_base).rstrip("/")
    deploy["verify_ssl"] = bool(verify_ssl)
    if operator_urls:
        urls = {k: str(v) for k, v in operator_urls.items() if v}
        if urls:
            deploy["operator_urls"] = urls
    dim["deploy"] = deploy
    rr["dim"] = dim
    out = dict(base)
    out[BEAMPIPE_RUN_RECORD_KEY] = rr
    return out


def _normalize_dim_graph_summary(graph: dict[str, Any] | None) -> dict[str, Any] | None:
    if not graph:
        return None
    out: dict[str, Any] = {}
    st = graph.get("status")
    if isinstance(st, str) and st:
        out["status"] = st
    count = graph.get("error_drop_count")
    if count is not None:
        try:
            out["error_drop_count"] = int(count)
        except (TypeError, ValueError):
            pass
    uids = graph.get("error_drop_uids")
    if isinstance(uids, list) and uids:
        clean = [str(u) for u in uids if u is not None][: _GRAPH_ERROR_UIDS_MAX]
        if clean:
            out["error_drop_uids"] = clean
    return out or None


def merge_dim_poll_into_manifest(
    existing_manifest: dict[str, Any] | None,
    *,
    session_id: str,
    session_state: str,
    http_status: int | None,
    record_terminal: bool,
    terminal_ledger_status: str | None,
    error: str | None = None,
    error_drops_count: int | None = None,
    graph: dict[str, Any] | None = None,
) -> dict[str, Any]:
    base: dict[str, Any] = dict(existing_manifest) if isinstance(existing_manifest, dict) else {}
    rr: dict[str, Any] = dict(base.get(BEAMPIPE_RUN_RECORD_KEY) or {})
    dim: dict[str, Any] = dict(rr.get("dim") or {})
    dim.pop("observations", None)
    dim["session_id"] = str(session_id)
    obs: dict[str, Any] = {
        "session_state": str(session_state),
        "http_status": http_status,
        "observed_at": _now_iso_z(),
    }
    dim["last_observation"] = obs

    graph_summary: dict[str, Any] | None = None
    if graph is not None:
        graph_summary = _normalize_dim_graph_summary(graph)
        if graph_summary is not None:
            dim["graph"] = graph_summary

    if record_terminal and terminal_ledger_status:
        term: dict[str, Any] = {
            **obs,
            "ledger_status": terminal_ledger_status,
            "session_id": str(session_id),
        }
        if error is not None:
            term["error"] = _trim_raw(error) or error[:_RAW_LINE_MAX]
        if error_drops_count is not None:
            term["error_drops_count"] = int(error_drops_count)
        if graph_summary is not None:
            term["graph"] = dict(graph_summary)
        existing_terminal = dim.get("terminal")
        frozen = (
            isinstance(existing_terminal, dict)
            and existing_terminal.get("ledger_status") == terminal_ledger_status
            and str(existing_terminal.get("session_id") or "") == str(session_id)
        )
        if not frozen:
            dim["terminal"] = term

    rr["dim"] = dim
    out = dict(base)
    out[BEAMPIPE_RUN_RECORD_KEY] = rr
    return out


def merge_restate_slurm_completion_timeout_into_manifest(
    existing_manifest: dict[str, Any] | None,
    *,
    error: str,
) -> dict[str, Any]:
    base: dict[str, Any] = dict(existing_manifest) if isinstance(existing_manifest, dict) else {}
    rr: dict[str, Any] = dict(base.get(BEAMPIPE_RUN_RECORD_KEY) or {})
    restate: dict[str, Any] = dict(rr.get("restate") or {})
    restate["slurm_completion"] = {
        "reason": "poll_timeout",
        "error": _trim_raw(error) or (error[:1024] + ("..." if len(error) > 1024 else "")),
        "observed_at": _now_iso_z(),
    }
    rr["restate"] = restate
    out = dict(base)
    out[BEAMPIPE_RUN_RECORD_KEY] = rr
    return out


def extract_beampipe_run_record(workflow_manifest: dict[str, Any] | None) -> dict[str, Any] | None:
    if not isinstance(workflow_manifest, dict):
        return None
    rr = workflow_manifest.get(BEAMPIPE_RUN_RECORD_KEY)
    return dict(rr) if isinstance(rr, dict) else None


def has_beampipe_run_record(workflow_manifest: dict[str, Any] | None) -> bool:
    return extract_beampipe_run_record(workflow_manifest) is not None


def preserve_run_record_into_manifest(
    new_manifest: dict[str, Any] | None,
    *,
    existing_manifest: dict[str, Any] | None,
) -> dict[str, Any]:
    new_dict: dict[str, Any] = dict(new_manifest) if isinstance(new_manifest, dict) else {}
    existing_rr = extract_beampipe_run_record(existing_manifest)
    if not existing_rr:
        return new_dict
    new_rr_raw = new_dict.get(BEAMPIPE_RUN_RECORD_KEY)
    new_rr: dict[str, Any] = dict(new_rr_raw) if isinstance(new_rr_raw, dict) else {}
    merged = {**existing_rr, **new_rr}
    new_dict[BEAMPIPE_RUN_RECORD_KEY] = merged
    return new_dict


def merge_execution_request_into_run_record(
    existing_manifest: dict[str, Any] | None,
    *,
    sources: list[Any] | None,
) -> dict[str, Any]:
    base: dict[str, Any] = dict(existing_manifest) if isinstance(existing_manifest, dict) else {}
    rr: dict[str, Any] = dict(base.get(BEAMPIPE_RUN_RECORD_KEY) or {})

    captured: list[dict[str, Any]] = []
    for spec in sources or []:
        if isinstance(spec, dict):
            sid = spec.get("source_identifier")
            if not sid:
                continue
            entry: dict[str, Any] = {"source_identifier": str(sid)}
            sbids = spec.get("sbids")
            if isinstance(sbids, list) and sbids:
                entry["sbids"] = [str(s) for s in sbids]
            captured.append(entry)
        else:
            sid = getattr(spec, "source_identifier", None)
            if not sid:
                continue
            entry = {"source_identifier": str(sid)}
            sbids = getattr(spec, "sbids", None)
            if isinstance(sbids, list) and sbids:
                entry["sbids"] = [str(s) for s in sbids]
            captured.append(entry)

    requested: dict[str, Any] = {
        "count": len(captured),
        "source_identifiers": [e["source_identifier"] for e in captured],
        "sources": captured,
        "captured_at": _now_iso_z(),
    }
    existing = rr.get("requested_sources")
    if isinstance(existing, dict):
        same = (
            existing.get("source_identifiers") == requested["source_identifiers"]
            and existing.get("count") == requested["count"]
        )
        if same:
            return base
    rr["requested_sources"] = requested
    out = dict(base)
    out[BEAMPIPE_RUN_RECORD_KEY] = rr
    return out

File: crates/beampipe-domain/src/run_record.rs
use chrono::{DateTime, Utc};
use serde_json::{json, Map, Value};

pub const BEAMPIPE_RUN_RECORD_KEY: &str = "beampipe_run_record";
const RAW_LINE_MAX: usize = 512;

fn now_iso_z() -> String {
    Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true)
}

fn trim_raw(raw: Option<&str>) -> Option<String> {
    let s = raw?.trim();
    if s.is_empty() {
        None
    } else if s.len() <= RAW_LINE_MAX {
        Some(s.to_string())
    } else {
        Some(format!("{}...", &s[..RAW_LINE_MAX]))
    }
}

fn object(v: Option<Value>) -> Map<String, Value> {
    match v {
        Some(Value::Object(map)) => map,
        _ => Map::new(),
    }
}

fn rr_mut(base: &mut Map<String, Value>) -> Map<String, Value> {
    object(base.remove(BEAMPIPE_RUN_RECORD_KEY))
}

pub fn merge_execution_request_into_run_record(
    existing: Option<Value>,
    sources: &[Value],
) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let captured: Vec<Value> = sources
        .iter()
        .filter_map(|spec| {
            let obj = spec.as_object()?;
            let sid = obj.get("source_identifier")?.as_str()?;
            let mut entry = Map::new();
            entry.insert("source_identifier".into(), Value::String(sid.to_string()));
            if let Some(Value::Array(sbids)) = obj.get("sbids") {
                if !sbids.is_empty() {
                    entry.insert(
                        "sbids".into(),
                        Value::Array(
                            sbids
                                .iter()
                                .map(|s| Value::String(s.to_string().trim_matches('"').to_string()))
                                .collect(),
                        ),
                    );
                }
            }
            Some(Value::Object(entry))
        })
        .collect();
    let ids: Vec<Value> = captured
        .iter()
        .filter_map(|v| v.get("source_identifier").cloned())
        .collect();
    rr.insert(
        "requested_sources".into(),
        json!({
            "count": captured.len(),
            "source_identifiers": ids,
            "sources": captured,
        }),
    );
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

pub fn merge_slurm_submit_into_manifest(
    existing: Option<Value>,
    session_id: &str,
    slurm_job_id: &str,
    composite_scheduler_job_id: &str,
    login_node: Option<&str>,
    remote_user: Option<&str>,
) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let mut slurm = object(rr.remove("slurm"));
    slurm.remove("observations");
    slurm.insert("session_id".into(), json!(session_id));
    slurm.insert("slurm_job_id".into(), json!(slurm_job_id));
    slurm.insert(
        "composite_scheduler_job_id".into(),
        json!(composite_scheduler_job_id),
    );
    slurm.insert("submitted_at".into(), json!(now_iso_z()));
    if let Some(v) = login_node.filter(|v| !v.trim().is_empty()) {
        slurm.insert("login_node".into(), json!(v.trim()));
    }
    if let Some(v) = remote_user.filter(|v| !v.trim().is_empty()) {
        slurm.insert("remote_user".into(), json!(v.trim()));
    }
    rr.insert("slurm".into(), Value::Object(slurm));
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

pub fn merge_scheduler_timeout_into_manifest(existing: Option<Value>, error: &str) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let mut scheduler = object(rr.remove("scheduler"));
    scheduler.insert(
        "slurm_completion".into(),
        json!({
            "reason": "poll_timeout",
            "error": trim_raw(Some(error)).unwrap_or_default(),
            "observed_at": now_iso_z(),
        }),
    );
    rr.insert("scheduler".into(), Value::Object(scheduler));
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

#[derive(Debug, Clone, Default)]
pub struct SlurmPollManifestOpts<'a> {
    pub exit_code: Option<i32>,
    pub remote_session_dir: Option<&'a str>,
    pub reason: Option<&'a str>,
    pub diagnostics: Option<Value>,
}

fn merge_slurm_paths(slurm: &mut Map<String, Value>, session_dir: &str) {
    let mut paths = object(slurm.remove("paths"));
    paths.insert("session_dir".into(), json!(session_dir.trim()));
    slurm.insert("paths".into(), Value::Object(paths));
}

#[allow(clippy::too_many_arguments)]
pub fn merge_slurm_poll_into_manifest(
    existing: Option<Value>,
    composite_scheduler_job_id: &str,
    slurm_job_id: &str,
    state: &str,
    source: &str,
    raw_line: Option<&str>,
    record_terminal: bool,
    terminal_ledger_status: Option<&str>,
    opts: SlurmPollManifestOpts<'_>,
) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let mut slurm = object(rr.remove("slurm"));
    slurm.remove("observations");
    slurm.insert("slurm_job_id".into(), json!(slurm_job_id));
    slurm.insert(
        "composite_scheduler_job_id".into(),
        json!(composite_scheduler_job_id),
    );
    if let Some(dir) = opts.remote_session_dir.filter(|d| !d.trim().is_empty()) {
        merge_slurm_paths(&mut slurm, dir);
    }
    let mut obs = Map::new();
    obs.insert("state".into(), json!(state));
    obs.insert("source".into(), json!(source));
    obs.insert("observed_at".into(), json!(now_iso_z()));
    if let Some(code) = opts.exit_code {
        obs.insert("exit_code".into(), json!(code));
    }
    if let Some(raw) = trim_raw(raw_line) {
        obs.insert("raw".into(), json!(raw));
    }
    slurm.insert("last_observation".into(), Value::Object(obs.clone()));
    if record_terminal {
        if let Some(status) = terminal_ledger_status {
            let comp = composite_scheduler_job_id;
            let frozen = slurm
                .get("terminal")
                .and_then(Value::as_object)
                .is_some_and(|t| {
                    t.get("ledger_status").and_then(Value::as_str) == Some(status)
                        && t.get("composite_scheduler_job_id").and_then(Value::as_str) == Some(comp)
                });
            if !frozen {
                let mut term = obs.clone();
                term.insert("ledger_status".into(), json!(status));
                term.insert(
                    "composite_scheduler_job_id".into(),
                    json!(composite_scheduler_job_id),
                );
                term.insert("slurm_job_id_ref".into(), json!(slurm_job_id));
                if let Some(reason) = opts.reason.filter(|r| !r.is_empty()) {
                    term.insert("reason".into(), json!(reason));
                }
                if let Some(diag) = opts.diagnostics.filter(|d| d.is_object()) {
                    term.insert("diagnostics".into(), diag);
                }
                slurm.insert("terminal".into(), Value::Object(term));
            }
        }
    }
    rr.insert("slurm".into(), Value::Object(slurm));
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

pub fn merge_dim_deploy_into_manifest(
    existing: Option<Value>,
    session_id: &str,
    dim_rest_base: &str,
    verify_ssl: bool,
    operator_urls: Option<std::collections::HashMap<String, String>>,
) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let mut dim = object(rr.remove("dim"));
    dim.remove("observations");
    dim.insert("session_id".into(), json!(session_id));
    let mut deploy = object(dim.remove("deploy"));
    deploy.insert("deployed_at".into(), json!(now_iso_z()));
    deploy.insert(
        "dim_rest_base".into(),
        json!(dim_rest_base.trim_end_matches('/')),
    );
    deploy.insert("verify_ssl".into(), json!(verify_ssl));
    if let Some(urls) = operator_urls {
        let filtered: Map<String, Value> = urls
            .into_iter()
            .filter(|(_, v)| !v.trim().is_empty())
            .map(|(k, v)| (k, json!(v)))
            .collect();
        if !filtered.is_empty() {
            deploy.insert("operator_urls".into(), Value::Object(filtered));
        }
    }
    dim.insert("deploy".into(), Value::Object(deploy));
    rr.insert("dim".into(), Value::Object(dim));
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

pub fn merge_dim_poll_into_manifest(
    existing: Option<Value>,
    session_id: &str,
    session_state: &str,
    record_terminal: bool,
    terminal_ledger_status: Option<&str>,
    error: Option<&str>,
    error_drops_count: Option<i64>,
) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let mut dim = object(rr.remove("dim"));
    dim.remove("observations");
    dim.insert("session_id".into(), json!(session_id));
    let mut obs = Map::new();
    obs.insert("session_state".into(), json!(session_state));
    obs.insert("observed_at".into(), json!(now_iso_z()));
    dim.insert("last_observation".into(), Value::Object(obs.clone()));
    if record_terminal {
        if let Some(status) = terminal_ledger_status {
            let mut term = obs;
            term.insert("ledger_status".into(), json!(status));
            term.insert("session_id".into(), json!(session_id));
            if let Some(err) = error.and_then(|e| trim_raw(Some(e))) {
                term.insert("error".into(), json!(err));
            }
            if let Some(count) = error_drops_count {
                term.insert("error_drops_count".into(), json!(count));
            }
            dim.insert("terminal".into(), Value::Object(term));
        }
    }
    rr.insert("dim".into(), Value::Object(dim));
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

/// Operator-facing summary of recent scheduler phases from run record.
pub fn summarize_run_record_phases(run_record: Option<&Value>) -> Value {
    let Some(rr) = run_record.and_then(|v| v.as_object()) else {
        return json!({});
    };
    let mut out = Map::new();
    for key in ["slurm", "dim", "scheduler", "slurm_poll", "dim_poll"] {
        if let Some(obj) = rr.get(key).and_then(Value::as_object) {
            let mut phase = Map::new();
            if let Some(t) = obj.get("started_at").or(obj.get("submitted_at")) {
                phase.insert("last_at".into(), t.clone());
            }
            if let Some(s) = obj.get("state").or(obj.get("terminal")) {
                phase.insert("state".into(), s.clone());
            }
            if let Some(r) = obj.get("round") {
                phase.insert("round".into(), r.clone());
            }
            if !phase.is_empty() {
                out.insert(key.into(), Value::Object(phase));
            }
        }
    }
    if let Some(req) = rr.get("requested_sources") {
        out.insert("requested_sources".into(), req.clone());
    }
    Value::Object(out)
}

pub fn extract_beampipe_run_record(workflow_manifest: &Value) -> Option<Value> {
    workflow_manifest
        .get(BEAMPIPE_RUN_RECORD_KEY)
        .cloned()
        .filter(|v| v.is_object())
}

pub fn parse_observed_at(v: &Value) -> Option<DateTime<Utc>> {
    v.as_str()
        .and_then(|s| DateTime::parse_from_rfc3339(s).ok())
        .map(|dt| dt.with_timezone(&Utc))
}

/// Poll round counter for batched `slurm_poll_tick` (stored under `beampipe_run_record.slurm_poll`).
pub fn slurm_poll_round_from_manifest(existing: Option<&Value>) -> i64 {
    existing
        .and_then(|m| m.get(BEAMPIPE_RUN_RECORD_KEY))
        .and_then(|rr| rr.get("slurm_poll"))
        .and_then(|p| p.get("round"))
        .and_then(serde_json::Value::as_i64)
        .unwrap_or(0)
}

pub fn merge_slurm_poll_tick_round(existing: Option<Value>, round: i64) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let mut tick = object(rr.remove("slurm_poll"));
    tick.insert("round".into(), json!(round));
    if !tick.contains_key("started_at") {
        tick.insert("started_at".into(), json!(now_iso_z()));
    }
    rr.insert("slurm_poll".into(), Value::Object(tick));
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

/// Poll round counter for batched `dim_poll_tick` (stored under `beampipe_run_record.dim_poll`).
pub fn dim_poll_round_from_manifest(existing: Option<&Value>) -> i64 {
    existing
        .and_then(|m| m.get(BEAMPIPE_RUN_RECORD_KEY))
        .and_then(|rr| rr.get("dim_poll"))
        .and_then(|p| p.get("round"))
        .and_then(serde_json::Value::as_i64)
        .unwrap_or(0)
}

pub fn merge_dim_poll_tick_round(existing: Option<Value>, round: i64) -> Value {
    let mut base = object(existing);
    let mut rr = rr_mut(&mut base);
    let mut tick = object(rr.remove("dim_poll"));
    tick.insert("round".into(), json!(round));
    if !tick.contains_key("started_at") {
        tick.insert("started_at".into(), json!(now_iso_z()));
    }
    rr.insert("dim_poll".into(), Value::Object(tick));
    base.insert(BEAMPIPE_RUN_RECORD_KEY.into(), Value::Object(rr));
    Value::Object(base)
}

pub fn dim_logs_url(dim_base: &str, session_id: &str) -> String {
    let base = dim_base.trim_end_matches('/');
    format!("{base}/api/sessions/{session_id}/logs")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn requested_sources_are_captured() {
        let out = merge_execution_request_into_run_record(
            None,
            &[json!({"source_identifier": "HIPASSJ1318-21", "sbids": ["1"]})],
        );
        assert_eq!(
            out["beampipe_run_record"]["requested_sources"]["source_identifiers"][0],
            "HIPASSJ1318-21"
        );
    }

    #[test]
    fn slurm_poll_terminal_freeze_preserves_first_terminal() {
        let first = merge_slurm_poll_into_manifest(
            None,
            "sid:1|/dir",
            "1",
            "COMPLETED",
            "sacct",
            Some("COMPLETED|0:0"),
            true,
            Some("completed"),
            SlurmPollManifestOpts {
                exit_code: Some(0),
                remote_session_dir: Some("/dir"),
                reason: None,
                diagnostics: None,
            },
        );
        let second = merge_slurm_poll_into_manifest(
            Some(first),
            "sid:1|/dir",
            "1",
            "COMPLETING",
            "sacct",
            Some("COMPLETING"),
            true,
            Some("completed"),
            SlurmPollManifestOpts::default(),
        );
        assert_eq!(
            second["beampipe_run_record"]["slurm"]["terminal"]["state"],
            "COMPLETED"
        );
        assert_eq!(
            second["beampipe_run_record"]["slurm"]["terminal"]["ledger_status"],
            "completed"
        );
    }

    #[test]
    fn slurm_poll_tick_round_is_stored() {
        let out = merge_slurm_poll_tick_round(None, 3);
        assert_eq!(out["beampipe_run_record"]["slurm_poll"]["round"], 3);
        assert!(out["beampipe_run_record"]["slurm_poll"]["started_at"].is_string());
    }

    #[test]
    fn slurm_submit_strips_legacy_observations() {
        let existing = json!({"beampipe_run_record": {"slurm": {"observations": []}}});
        let out = merge_slurm_submit_into_manifest(Some(existing), "s", "1", "s:1", None, None);
        assert!(out["beampipe_run_record"]["slurm"]
            .get("observations")
            .is_none());
        assert_eq!(out["beampipe_run_record"]["slurm"]["slurm_job_id"], "1");
    }
}
